doc/neps/nep-0047-array-api-standard.rst

```diff
@@ -0,0 +1,534 @@
+protocol. Both work similarly, letting the "producer" describe how the data
+is laid out in memory so the "consumer" can construct its own kind of array
+with a view on that data.
+
+DLPack works in a very similar way. The main reasons to prefer DLPack over
+the options already present in NumPy are:
+
+1. DLPack is the only protocol with device support (e.g., GPUs using CUDA or
+   ROCm drivers, or OpenCL devices). NumPy is CPU-only, but other array
+   libraries are not. Having one protocol per device isn't tenable, hence
+   device support is a must.
+2. Widespread support. DLPack has the widest adoption of all protocols, only
+   NumPy is missing support. And the experiences of other libraries with it
+   are positive. This contrasts with the protocols NumPy does support, which
+   are used very little - when other libraries want to interoperate with
+   NumPy, they typically use the (more limited, and NumPy-specific)
+   ``__array__`` protocol.
+
+Adding support for DLPack to NumPy entails:
+
+- Adding a ``ndarray.__dlpack__`` method
+- Adding a ``from_dlpack`` function, which takes as input an object
+  supporting ``__dlpack__``, and returns an ``ndarray``.
+
+DLPack is currently a ~200 LoC header, and is meant to be included directly, so
+no external dependency is needed. Implementation should be straightforward.
+
+
+Syntax for device support
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+NumPy itself is CPU-only, so it clearly doesn't have a need for device support.
+However, other libraries (e.g. TensorFlow, PyTorch, JAX, MXNet) support
+multiple types of devices: CPU, GPU, TPU, and more exotic hardware.
+To write portable code on systems with multiple devices, it's often necessary
+to create new arrays on the same device as some other array, or check that
+two arrays live on the same device. Hence syntax for that is needed.
+
+The array object will have a ``.device`` attribute which enables comparing
+devices of different arrays (they only should compare equal if both arrays are
+from the same library and it's the same hardware device). Furthermore,
+``device=`` keywords in array creation functions are needed. For example::
+
+    def empty(shape: Union[int, Tuple[int, ...]], /, *,
+              dtype: Optional[dtype] = None,
+              device: Optional[device] = None) -> array:
+        """
+        Array API compatible wrapper for :py:func:`np.empty <numpy.empty>`.
+        """
+        return np.empty(shape, dtype=dtype, device=device)
+
+The implementation for NumPy may be as simple as setting the device attribute to
+the string ``'cpu'`` and raising an exception if array creation functions
+encounter any other value.
+
+
+Dtypes and casting rules
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The supported dtypes in this namespace are boolean, 8/16/32/64-bit signed and
+unsigned integer, and 32/64-bit floating-point dtypes. These will be added to
+the namespace as dtype literals with the expected names (e.g., ``bool``,
+``uint16``, ``float64``).
```

Comment:
It may be worth clarifying what the exact type of "dtype literal" objects will be for NumPy, even if the API standard leaves this as ambiguous.

For example, are these Python `type` objects? Is there any subclass/subtyping relationship between them?

Reply:
I'd suggest just aliasing them to `float64` et al. in the main namespace. And having a private validation function for `dtype=` to ensure users pass in one of those aliases and not a format string for example.

Reply:
Although on reflection, I think we want this to work:
```
numpy.array_api.some_func(..., dtype=x.dtype)
```
Right now:
```
>>> x.dtype
dtype('int64')
>>> x.dtype is np.int64
False
>>> x.dtype is np.dtype(np.int64)
True
```
Making them `np.dtype` instances could make sense. Or some singleton class - that would not expose all the unnecessary dtype class attributes, but it'd be annoying wrapping.

For now I'll add: `np.array_api.float64 = np.dtype(np.float64)`. If you see a problem with that or have a better idea, please let me know.

Reply:
> Is there any subclass/subtyping relationship between them?

That's a good question. I'd prefer to keep that as simple as possible. We should figure out if we need anything from the `np.is*` set of functions. There's an open numpy issue about these dtype functions being a mess, we should probably just link to https://github.com/numpy/numpy/issues/17325 and see if we can resolve that.

Using `isinstance` is bad, something like `np.issubdtype(x.dtype, np.integer)` is worse imho.

In the standard we have `isnan` and `isinf`, maybe we should add `isinteger`, `isfloating`.

Reply:
Or maybe it's now simple enough to do:
```
def isfloating(dtype):
    return dtype in (float32, float64)

def isinteger(dtype):
    return dtype in (uint8, unit16, ..., int64)
```
in user code. maybe something worth opening an issue about on the array-api repo.

Reply:
added a section on dtype objects under _Implementation_.